pkg/build/strategies/sti/postexecutorstep.go

```diff
@@ -0,0 +1,461 @@
+		if err := step.fs.Copy(src, dst); err != nil {
+			return fmt.Errorf("Couldn't copy file (%q -> %q): %v", src, dst, err)
+		}
+	}
+
+	image := step.builder.config.RuntimeImage
+	workDir, err := step.docker.GetImageWorkdir(image)
+	if err != nil {
+		return fmt.Errorf("Couldn't get working dir of %q image: %v", image, err)
+	}
+
+	user, err := step.docker.GetImageUser(image)
+	if err != nil {
+		return fmt.Errorf("Couldn't get user of %q image: %v", image, err)
+	}
+
+	commandBaseDir := workDir
+	if !useExternalAssembleScript {
+		// script already inside of the image
+		scriptsURL, err := step.docker.GetScriptsURL(image)
+		if err != nil {
+			return err
+		}
+		if len(scriptsURL) == 0 {
+			return fmt.Errorf("Couldn't determine scripts URL for image %q", image)
+		}
+		commandBaseDir = strings.TrimPrefix(scriptsURL, "image://")
+	}
+
+	cmd := fmt.Sprintf(
+		"while [ ! -f %q ]; do sleep 0.5; done; %s/%s; exit $?",
+		lastFileDstPath,
+		commandBaseDir,
+		api.AssembleRuntime,
+	)
+
+	opts := dockerpkg.RunContainerOptions{
+		Image:           image,
+		PullImage:       false, // The PullImage is false because we've already pulled the image
+		CommandExplicit: []string{"/bin/sh", "-c", cmd},
+		Stdout:          outWriter,
+		Stderr:          errWriter,
+		NetworkMode:     string(step.builder.config.DockerNetworkMode),
+		CGroupLimits:    step.builder.config.CGroupLimits,
+		CapDrop:         step.builder.config.DropCapabilities,
+		PostExec:        step.builder.postExecutor,
+		Env:             step.builder.env,
+		User:            user,
+	}
+
+	opts.OnStart = func(containerID string) error {
+		setStandardPerms := func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			// Skip chmod for symlinks
+			if info.Mode()&os.ModeSymlink != 0 {
+				return nil
+			}
+			// file should be writable by owner (u=w) and readable by other users (a=r),
+			// executable bit should be left as is
+			mode := os.FileMode(0644)
+			if info.IsDir() || info.Mode()&syscall.S_IEXEC != 0 {
+				mode = 0755
+			}
+			return step.fs.Chmod(path, mode)
+		}
+
+		glog.V(5).Infof("Uploading directory %q -> %q", artifactsDir, workDir)
+		if err := step.docker.UploadToContainerWithCallback(artifactsDir, workDir, containerID, setStandardPerms, true); err != nil {
+			return fmt.Errorf("Couldn't upload directory (%q -> %q) into container %s: %v", artifactsDir, workDir, containerID, err)
+		}
+
+		glog.V(5).Infof("Uploading file %q -> %q", lastFilePath, lastFileDstPath)
+		if err := step.docker.UploadToContainerWithCallback(lastFilePath, lastFileDstPath, containerID, setStandardPerms, true); err != nil {
+			return fmt.Errorf("Couldn't upload file (%q -> %q) into container %s: %v", lastFilePath, lastFileDstPath, containerID, err)
+		}
+
+		return err
+	}
+
+	go dockerpkg.StreamContainerIO(outReader, nil, func(message ...interface{}) {
+		glog.V(0).Infof("[runtime image stdout] %s", message...)
```

Comment:
>because the message will appear in between a header/footer that say

I'm afraid it could lead to a mess in case of, for example, running `mvn compile`.

(I'll test this case on Thursday.)